src/base/bittorrent/private/index.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2018  Eugene Shalygin <eugene.shalygin@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <libtorrent/version.hpp>
+#if LIBTORRENT_VERSION_NUM >= 10200
+#include <libtorrent/units.hpp>
+#endif
+
+namespace BitTorrent
+{
+#if LIBTORRENT_VERSION_NUM >= 10200
+
+    // these typedefs are needed until we drop C++11
+    using libt_file_index_t = libtorrent::file_index_t;
+    using libt_piece_index_t = libtorrent::piece_index_t;
+
+    inline constexpr libtorrent::file_index_t makeFileIndex(libtorrent::file_index_t::underlying_type index)
```

Comment:
It would be slightly longer names to use a template. However, instead of these functions, you could just construct `libt_file_index_t` objects directly (and in 1.1.x it would just construct an `int`)

Reply:
I agree that typedefs and functions are somewhat overlap. I used functions because I knew I can auto-type variables then. Now, with C++11 it does not work always and typedefs are required. I will eliminate the overlapping.

Reply:
This general concept is typically referred to as "strong types". It's become a very popular thing in the C++ community in the last 5 years or so (at least to talk about). Personally I've employed it pervasively at my current and last job, with great success.

First, to respond to the meaningless type conversions: if you store piece indices, file indices, download priorities, peer classes, etc. in variables of the specific type, you don't need the type conversion (and you gain the type-safety). The annoyance of the type conversions is a feature, to indicate that the parameter- or variable type should should be changed to match what you're passing in.

A counter-question could be: Why do you use an `int` to store a piece index? The vast majority of operations you can perform on an integer are not necessary and most likely bugs.

In my experience, once you embrace the type system fully and make sure all your types match, or at least maximises the times they match, you really gain a lot of confidence that your code fits together correctly. (I would argue that the [mars climate orbiter crash](https://en.wikipedia.org/wiki/Mars_Climate_Orbiter#Cause_of_failure) belongs in this category of errors). I also had a type related bug in code converting from internal file mode flags to OS specific ones, where an operator precedence mistake caused the flags to be mixed with each other.

Sometimes embracing the type system requires you to get creative, to make sure that your argument types match. For example, `std::vector<T>::operator[]` takes a `std::size_t` as the index. If you want to pass in a `piece_index_t`, what do you do? You can create a new template, that derives from `std::vector` but replaces all index types with a template parameter. (That's what I typically [do](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/aux_/vector.hpp#L43)). In fact, I've done this for so many containers (`std::array`, `std::deque`) that I generalised the concept of [wrapping a cointainer](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/aux_/container_wrapper.hpp), specifying the index type. I've also wrapped [`std::unique_ptr<>`](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/aux_/unique_ptr.hpp) (the version that holds an array) to allow specifying the index type. There are [other](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/aux_/typed_span.hpp#L45) [examples](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/bitfield.hpp#L268) too.

Apart from interacting with containers, I think I/O is the other major hurdle to use specific types.
The modern ways of printing has a much better story here, for instance implementing a templated `operator<<` and `operator>>` for iostream is pretty simple. Or a hook for `fmtlib`.

Granted, if you stick to `printf()` you will need to cast. But I don't ascribe a lot of value to be compatible with something that's so unsafe to begin with anyway (it could be seen as a feature, to discourage use of `printf()`, and ellipsis in general).

The third case I can think of is where you need some specific "special" operation. For instance find the piece index in the middle between two other indices, or find the range of pieces that represent a certain number of bytes. My approach to these kinds of operations is to wrap them in functions that have a type-safe interface, and then do the casting internally.

I'm genuinely interested if you have cases where you're forced to cast that doesn't fit one of the categories I outlined (or if there's a case that's common but doesn't have a good solution for using strong types).

I also recognise that converting an existing program to use strong types is quite a daunting task, and that it might not be super exciting to do.

Reply:
@arvidn, thank you for detailed description. I'm just a hobbyist and I don't have enough experience to understand some concepts at once (I see how it's used here and it raises questions for me). Although I think the code would look nicer if we didn't have to support both libtorrent versions.
Now I have food for thought. If I have more specific comments, I'll call you.